feat searchlib: add an option to prefetch tensors in HNSW index search

itrofimow opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

----

TLDR: ~x2 speedup of HNSW index by `__builtin_prefetch`-ing tensors

Hi!
I was benchmarking my Vespa setup the other day, and decided to get a flamegraph of what proton is doing:
![trunk](https://github.com/user-attachments/assets/b6f2b563-d21b-4a1f-b0ef-164c2a2cbccb)
As expected, a lot of cpu cycles are being spent in HNSW index doing some HNSW things, one of such things being [`binary_hamming_distance`](https://github.com/vespa-engine/vespa/blob/7691c8ee3ed5533faf2ebff51aa9b4d92a0bddfc/vespalib/src/vespa/vespalib/util/binary_hamming_distance.cpp#L14). I decided to give it a closer look and soon realized that in my setup (gcc 14.2, aarch64) gcc fails miserably to produce unrolled/vectorized version of the code, and that it could probably be improved by either SimSIMD or hand-written intrinsincs. Inspired, I implemented a version of `binary_hamming_distance` that beats current code ~x1.8 in micro-benchmarks, deployed it, benchmarked... and saw no difference at all, none, zero.

That confused me, so I decided to give another function a look: [`get_num_subspaces_and_flag`](https://github.com/vespa-engine/vespa/blob/7691c8ee3ed5533faf2ebff51aa9b4d92a0bddfc/searchlib/src/vespa/searchlib/tensor/tensor_buffer_operations.cpp#L54). Well, that's a one (3 with asserts) instruction function, huh? Could it be the overhead of it being in another TU and actually requiring a `call`? - well.. unlikely. That confused me even more ~~and I went on to check if my perf is working ok~~

----

At some point I realized that I have a HNSW with ~1B of `tensor<int8>(d1[128])`, which is a lot of memory with very unpredictable access pattern, and that 1 instruction `get_num_subspaces_and_flag` function is actually a memory load, so I built a flamegraph for last-level-cache misses (`perf record -e LLC-load-misses`), and suddenly everything made sense:
![trunk_llc_misses](https://github.com/user-attachments/assets/7499cf0a-a4e5-4671-a5f5-3a678001a29a)
As one can see, flamegraphs for cpu-cycles and LLC-load-misses look basically the same for HNSW index.

Looking closer at perf data for LLC I concluded that
* LLC miss in `search::tensor::SerializedFastValueAttribute::get_vector` is coming from here https://github.com/vespa-engine/vespa/blob/7691c8ee3ed5533faf2ebff51aa9b4d92a0bddfc/searchlib/src/vespa/searchlib/tensor/serialized_fast_value_attribute.cpp#L47, specifically from [this load](https://github.com/vespa-engine/vespa/blob/7691c8ee3ed5533faf2ebff51aa9b4d92a0bddfc/searchlib/src/vespa/searchlib/tensor/tensor_attribute.h#L45)
* LLC miss in `get_num_subspaces_and_flag` is a load from tensor buffer
* LLC misses in `binary_hamming_distance`  are also loads from tensor buffer

----

The good thing about HNSW memory access pattern is that although it's very hard for hardware to predict and prefetch, we know exactly what memory we will have to access: [given a vertex in the graph, check all its neighbors](https://github.com/vespa-engine/vespa/blob/7691c8ee3ed5533faf2ebff51aa9b4d92a0bddfc/searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp#L395) , thus we can rearrange how we walk the neighbors in such a way that with some hints to hardware there would be way less misses:
1. for every neighbor, prefetch the memory in `TensorAttribute::_refVector`
2. for every neighbor, prefetch its tensor in `TensorBufferOperations` (which requires a load from `TensorAttribute::_refVector`, but hopefully at this point the memory would already be brought to caches)
3. for every neighbor, do what we currently do (and hopefully the tensor memory would already be in caches)

That's a lot of "hopefully", but that's just how prefetching hints work, and turns out they work wonders: with the patch applied flamegraphs for proton look like this
![patch](https://github.com/user-attachments/assets/d314ca6c-52ee-4288-b84c-8c1c33063ea3)

with the LLC-load-misses flamegraph looking almost the same as it did, as expected (LLC-misses from prefetching are still there, but now they are async and don't stall us that much)
![patch_llc_misses](https://github.com/user-attachments/assets/f2140ab4-8c3c-4874-9074-177738865191)

Comparing trunk cpu-cycles flamegraph with patch cpu-cycles flamegraph it looks like there are ~x2 less cycles spent in HSNW index, which amount to ~10% of total cpu cycles spent, and that matches with CPU usage/timings I'm observing when benchmarking.

All benchmarks were conducted at commit b60aa0ded117458f271df411bbfc23d838320afc, ~1B of `tensor<int8>(d1[128])`, AWS Graviton4

itrofimow commented:
Hi @boeker ! I see you've committed a plenty into hnsw_index.cpp recently, would you be able to give this PR a look?

vekterli commented:
Sorry about the delay, we've been rather busy 😓 Hoping we'll get around to it this week.

My gut feeling is that we may want to template the search layer functions (filtered and unfiltered) on some kind of prefetching policy and then have a runtime decision on whether we want to actually prefetch anything, based on config and/or the size of the graph. In my anecdotal experience, explicitly prefetching often makes the performance go _down_ if enough stuff is already present in the cache hierarchy, which may be the case for small graphs. But at some point the curves will intersect and prefetching should present an increasingly visible gain.

Could also be an interesting experiment to see if prefetching into only caches > L1 would be beneficial. L1D is comparatively tiny, so when prefetching many vectors we may (emphasis: _may_—needs benchmarking!) risk evicting useful stuff from it that we'll end up needing before we actually get around to using the vectors themselves.

vekterli commented on searchlib/src/vespa/searchlib/tensor/tensor_buffer_store.h:
```diff
@@ -72,6 +72,13 @@ class TensorBufferStore : public TensorStore
+
+        for (std::size_t i = 0; i < buf.size(); i += 64) {
+            __builtin_prefetch(buf.data() + i);
+        }
```
I'm a bit concerned that this ends prefetching _everything_ across all subspaces, which can pollute the caches with stuff that will not be used (especially L1d)...

The first cache line of `buf` holds the subspace header, so maybe that should be prefetched first, then prefetch the actual subspace afterwards.

vekterli reviewed (commented):
Sorry for the delay, lots of stuff popping up... 👀

As I alluded to in the `TensorBufferStore` code, I have a concern that we risk polluting the caches with the current approach when there are many subspaces. Ideally we should start out with tensors stored in `DenseTensorStore` to avoid this risk, since dense tensors do not have multiple subspaces.

Re: my earlier comment/question, it is not entirely clear to me why your seemingly dense tensors ended up instantiating a `SerializedFastValueAttribute`, so that would be good to figure out first.

itrofimow commented:
Sorry for the late reply, I also got consumed by other things.

Regarding the tensor type: it actually is defined as `type tensor<int8>(d0{},d1[128])`; I've omitted the `d0{}` part due to thinking it is not significant. My bad, sorry for the confusion.

I am also not sure that prefetching the whole tensor data is a good thing to do due to exactly the same concerns you outlined above. Initially I tried prefetching just the first 4 cache-lines, but that felt arbitrary, so I ended up with the current approach with no noticeable difference.
Regarding prefetching the tensor header first and then prefetching the subspace needed: I'm not sure there would be enough time for memory subsystem to actually bring the header in before it's accessed in an attempt to prefetch the subspace.

Unfortunately, I won't be able to easily measure this change with dense tensors, as there aren't any in my setup.
Following up on the size of the sparse tensor prefetch, I think it could be okay to do the full prefetch as long as prefetching is allowed by the policy we discussed above, what do you think?

Commit:
[searchlib] prefetch tensors in HNSW index search

itrofimow commented:
I've addressed your inline comments and implemented a simple on/off policy for the prefetching.

Please forgive the force-push: I'm upstreaming these changes from a fork that considerably lags behind, and rebasing on top of current master turned out to be non-trivial due to recent changes.

Commit:
add the java part of prefetch-tensors ranking feature

itrofimow commented:
Thanks @boeker ! Basically, me:

<img width="700" height="720" alt="image" src="https://github.com/user-attachments/assets/25494128-fd5f-4dde-87d0-2636f1a0e3bc" />

vekterli commented on searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp:
```diff
@@ -288,6 +288,16 @@ calc_distance_helper(const BoundDistanceFunction &df, vespalib::eval::TypedCells
+void prefetch_vector(const DocVectorAccess& vectors, uint32_t docid)
+{
+    vectors.prefetch_vector(docid);
+}
```
Should these just be `vectors.prefetch_...` directly in the caller instead of going via forwarding freestanding functions?

vekterli reviewed (commented):
I think this looks good, just some minor nitpicks.

(Would also be great if @arnej27959 or @toregge could have a quick look at the code that touches the tensor/data stores :eyes:)

We've had an internal discussion regarding the added rank profile property and consensus is that it makes sense to add as-is. However, we most likely want to do some additional work around this before we feel it's ready to be an officially documented feature. Off the top of my head:

* We probably want a client query property that mirrors the rank profile property which can override prefetching on a per query basis.
* Prefetching of sparse tensors should be subspace-aware.
* We should unify prefetching between `search_layer_helper` and `search_layer_filter_first_helper`.
* Since updates to the graph also involves a search, we should look into if/how we can expose _feed-time_ prefetching as well, not just _query-time_.
* We should set up and do our own large-scale (100M-1B) HNSW index experiments so that we have an environment where we can measure the impact of toggling prefetching.
* We need to understand how prefetching interacts with paged tensors, both when tensors fit in the buffer cache and when the combined tensor footprint is >> main memory size.

Ideally we'd get this merged very soon to start getting perf test results (this will also tell us if we need to make the neighbor prefetching conditional for small graphs), but I have a heap of unspent vacation days that have accumulated and are bursting at the seams, so I can't really follow up this until after new year's :sweat_smile: :santa: Someone else™ would need to follow this up until then.

Commit:
Merge remote-tracking branch 'origin/master' into hnsw_index_prefetch_tensors

Commit:
cr fixes: fix typos and add some comments

Commit:
oops

Commit:
oops, typo fix

Commit:
Merge remote-tracking branch 'itrofimow/master' into HEAD

Commit:
fix the build after rebase

vekterli reviewed (approved):
:+1: LGTM! We've had some upstream service provider issues in our build systems, so that's why I haven't pulled this one in yet (I want a couple of recent and stable internal performance test runs prior to merging). Will do it once things stabilize, presumably either today or tomorrow.
